src/jllvm/class/ByteCodeIterator.hpp

```diff
@@ -216,30 +216,319 @@ enum class OpCodes : std::uint8_t
+    ALoad = 0x19,
+    AStore = 0x3a,
+    DLoad = 0x18,
+    DStore = 0x39,
+    FLoad = 0x17,
+    FStore = 0x38,
+    ILoad = 0x15,
+    IStore = 0x36,
+    LLoad = 0x16,
+    LStore = 0x37,
+    Ret = 0xa9,
+};
+
+enum class PoolIndexedOpCodes : std::uint8_t
+{
+    LDC = 0x12, // TODO only 1 byte
+    ANewArray = 0xbd,
+    CheckCast = 0xc0,
+    GetField = 0xb4,
+    GetStatic = 0xb2,
+    InstanceOf = 0xc1,
+    InvokeDynamic = 0xba,   // TODO check padding 0 + 0
+    InvokeInterface = 0xb9, // TODO check padding not 0 + 0
+    InvokeSpecial = 0xb7,
+    InvokeStatic = 0xb8,
+    InvokeVirtual = 0xb6,
+    LDCW = 0x13,
+    LDC2W = 0x14,
+    New = 0xbb,
+    PutField = 0xb5,
+    PutStatic = 0xb3,
+};
+
+enum class BranchOffsetOpCodes : std::uint8_t
+{
+    Goto = 0xa7,
+    GotoW = 0xc8,
+    IfACmpEq = 0xa5,
+    IfACmpNe = 0xa6,
+    IfICmpEq = 0x9f,
+    IfICmpNe = 0xa0,
+    IfICmpLt = 0xa1,
+    IfICmpGe = 0xa2,
+    IfICmpGt = 0xa3,
+    IfICmpLe = 0xa4,
+    IfEq = 0x99,
+    IfNe = 0x9a,
+    IfLt = 0x9b,
+    IfGe = 0x9c,
+    IfGt = 0x9d,
+    IfLe = 0x9e,
+    IfNonNull = 0xc7,
+    IfNull = 0xc6,
+    JSR = 0xa8,
+    JSRw = 0xc9,
+};
+
+enum class SpecialOpCodes : std::uint8_t
+{
+    BIPush = 0x10,
+    NewArray = 0xbc,
+    IInc = 0x84,
+    SIPush = 0x11,
+    LookupSwitch = 0xab,
+    MultiANewArray = 0xc5,
+    TableSwitch = 0xaa,
+    Wide = 0xc4,
+};
+
+struct LocalIndexedOp
+{
+    LocalIndexedOpCodes opCode{};
+
+    std::uint8_t index{};
+};
+
+struct PoolIndexedOp
+{
+    PoolIndexedOpCodes opCode{};
+
+    std::uint16_t index{};
+};
+
+struct BranchOffsetOp
+{
+    BranchOffsetOpCodes opCode{};
+
+    std::int32_t target{};
+
     std::size_t offset{};
-    /// Reference to any variable-length extra data this operation contains.
-    /// Does not include the identifying byte of the instruction.
-    llvm::ArrayRef<char> data;
 };
 
+struct BIPushOp
+{
+    std::int8_t value{};
+};
+
+struct NewArrayOp
+{
+    enum class ArrayType : std::uint8_t
+    {
+        TBoolean = 4,
+        TChar = 5,
+        TFloat = 6,
+        TDouble = 7,
+        TByte = 8,
+        TShort = 9,
+        TInt = 10,
+        TLong = 11
+    };
+
+    ArrayType atype{};
+};
+
+struct IIncOp
+{
+    std::uint8_t index{};
+
+    std::int8_t byte{};
+};
+
+struct SIPushOp
+{
+    std::int16_t value{};
+};
+
+struct LookupSwitchOp
+{
+};
+
+struct MultiANewArrayOp
+{
+};
+
+struct TableSwitchOp
+{
+};
+
+struct WideOp
+{
+};
+
+using ByteCodeOp = std::variant<SingletonOp, LocalIndexedOp, PoolIndexedOp, BranchOffsetOp, BIPushOp, NewArrayOp,
+                                IIncOp, SIPushOp, LookupSwitchOp, MultiANewArrayOp, TableSwitchOp, WideOp>;
+
 class ByteCodeIterator : public llvm::iterator_facade_base<ByteCodeIterator, std::forward_iterator_tag, ByteCodeOp,
                                                            std::ptrdiff_t, ByteCodeOp, ByteCodeOp>
 {
-    const char* m_current = nullptr;
+    llvm::ArrayRef<char> m_current;
```

Comment:
Can we keep this `const char*`? Using `llvm::ArrayRef` is not required here since we don't need to know the end from one instance of an iterator and it uses 8 extra bytes for no reason.